README.md
# [Conda Cheatsheet](https://marketplace.visualstudio.com/items?itemName=jojoee.conda-cheatsheet)
A Visual Studio Code extension that lets you open Conda cheatsheet inside the editor.

[![Version](https://vsmarketplacebadge.apphb.com/version/jojoee.conda-cheatsheet.svg)](https://marketplace.visualstudio.com/items?itemName=jojoee.conda-cheatsheet)
[![Installs](https://vsmarketplacebadge.apphb.com/installs-short/jojoee.conda-cheatsheet.svg)](https://marketplace.visualstudio.com/items?itemName=jojoee.conda-cheatsheet)
[![Rating](https://vsmarketplacebadge.apphb.com/rating-short/jojoee.conda-cheatsheet.svg)](https://marketplace.visualstudio.com/items?itemName=jojoee.conda-cheatsheet)

![Demo](./asset/screenshot/demo.gif)
![PDF](./asset/screenshot/pdf-768.png)
![Support Dark and Light Theme](./asset/screenshot/support-dark-and-light-theme-1024.png)

## Features
- `Conda Cheatsheet: PDF` to open PDF file inside editor
- `Conda Cheatsheet: Website` to open website inside editor
- `Conda Cheatsheet: Webview` to open website inside editor
  - Support both light and dark theme
  - Click on the content section for jump to selected section immediately
  - Copy and Paste command

## TODO
- [ ] Automated: download pdf => convert to image for `onCommand:extension.openPdf` command
- [ ] Automated: test
- [ ] Automated: watch and run `compileWebview.js` file
- [ ] Refactoring

## Demo
1. Basic usage + feature
- `alt + shift + P` to open available commands
- Select for open "Webview"
- Click to "Sharing Environments" section
- Hover on some command
- Click it for copy the command
- `alt + N` to create new file
- `alt + V` to pate the command
2. Support theme changing
3. Support other view e.g. PDF and website

## Development
```
node tool/compileWebview.js && npm run compile

npm run compile
npm run test
npm run deploy
vsce login jojoee
vsce publish
```

## Reference
- [Conda - Conda documentation](https://docs.conda.io/en/latest/)


CHANGELOG.md
# Change Log

All notable changes to the "conda-cheatsheet" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [1.2.8]
- Add Conda cheatsheet version 4.12
- Add Conda cheatsheet version 4.14

## [1.2.7]
- Downgrade required engine to 1.26.0

## [1.2.6]
- Config CI

## [1.2.5]
- Update screenshot files

## [1.2.4]
- Add screenshot files
- Update readme file

## [1.2.3]
- Using pre-compiled template instead of compiling it on-the-fly (for reducing published-extension size)

## [1.2.2]
- Fix and update icon.png
- Fix the plugin is not working

## [1.2.1]
- Add icon file
- Using `setStatusBarMessage` instead of `showInformationMessage` API
- Click the code section to copy it
- Update Webview css

## [1.2.0]
- Release `extension.openWebview` command

## [1.1.0]
- Release `extension.openWebsite` command

## [1.0.0]
- Initial release
- Release `extension.openPdf` command


src/test/suite/index.ts
import * as path from 'path'
import * as Mocha from 'mocha'
import * as glob from 'glob'

export function run (): Promise<void> {
  // Create the mocha test
  const mocha = new Mocha({
    ui: 'tdd'
  })
  mocha.useColors(true)

  const testsRoot = path.resolve(__dirname, '..')

  return new Promise((c, e) => {
    glob('**/**.test.js', { cwd: testsRoot }, (err, files) => {
      if (err) {
        return e(err)
      }

      // Add files to the test suite
      files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)))

      try {
        // Run the mocha test
        mocha.run(failures => {
          if (failures > 0) {
            e(new Error(`${failures} tests failed.`))
          } else {
            c()
          }
        })
      } catch (err) {
        console.error(err)
        e(err)
      }
    })
  })
}


src/content.ts
let sections = [{
  name: 'Getting Started',
  commands: [
    {
      desc: 'Verify Conda is installed, check version number',
      code: 'conda info'
    },
    {
      desc: 'Update Conda to the current version',
      code: 'conda update -n base conda'
    },
    {
      desc: 'Update all packages to teh latest version of Anaconda. Will install stable and compatible versions, not necessarily the very latest.',
      code: 'conda update anaconda'
    }
  ]
}, {
  name: 'Working with Environments',
  commands: [
    {
      desc: 'Create a new environment named ENVNAME with specific version of Python and packages installed.',
      code: 'conda create --name <ENVNAME> python=3.6 "PKG1>7.6" PKG2'
    },
    {
      desc: 'Activate a named Conda environment',
      code: 'conda activate ENVNAME'
    },
    {
      desc: 'Activate a Conda environment at particular location on disk',
      code: 'conda activate /path/to/environment-dir'
    },
    {
      desc: 'Deactivate current environment',
      code: 'conda deactivate'
    },
    {
      desc: 'List all packages and versions in the active environment',
      code: 'conda list'
    },
    {
      desc: 'List all packages and versions in the named environment',
      code: 'conda list --name ENVNAME'
    },
    {
      desc: 'List all revisions made within the active environment',
      code: 'conda list --revisions'
    },
    {
      desc: 'List all revisions made in a specified environment',
      code: 'conda list --name ENVNAME --revisions'
    },
    {
      desc: 'Delete an entire environment',
      code: 'conda remove --name ENVNAME --all'
    }
  ]
}, {
  name: 'Sharing Environments',
  commands: [
    {
      desc: 'Make an exact copy of an environment',
      code: 'conda create --clone ENVNAME --name NEWENV'
    },
    {
      desc: 'Export an environment to a YAML file that can be read on Windows, macOS, and Linux',
      code: 'conda env export --name ENVNAME > envname.yml'
    },
    {
      desc: 'Create an environment from YAML file',
      code: 'conda env create --file envname.yml'
    },
    {
      desc: 'Create an environment from the file named environment.yml in the current directory',
      code: 'conda env create'
    },
    {
      desc: 'Export an environment with exact package version for one OS',
      code: 'conda list --explicit > pkgs.txt'
    },
    {
      desc: 'Create an environment based on exact package versions',
      code: 'conda create --name NEWENV --file pkgs.txt'
    }
  ]
}, {
  name: 'Using Packages and Channels',
  commands: [
    {
      desc: 'Search for a package in currently configured channels with version range >=3.1.0, <3.2"',
      code: 'conda search PKGNAME=3.1 "PKGNAME [version=\'>=3.1.0,<3.2\']"'
    },
    {
      desc: 'Find a package on all channels using the Anaconda Client',
      code: 'anaconda search FUZZYNAME'
    },
    {
      desc: 'Install package from a specific channel',
      code: 'conda install conda-forge::PKGNAME'
    },
    {
      desc: 'Install a package by exact version number (3.1.4)',
      code: 'conda install PKGNAME==3.1.4'
    },
    {
      desc: 'Install one of the listed versions (OR)',
      code: 'conda install "PKGNAME[version=\'3.1.2|3.1.4\']"'
    },
    {
      desc: 'Install following several constraints (AND)',
      code: 'conda install "PKGNAME>2.5,<3.2"'
    },
    {
      desc: 'Add a channel to your Conda configuration',
      code: 'conda config --add channels CHANNELNAME'
    }
  ]
}, {
  name: 'Additional Useful Hints',
  commands: [
    {
      desc: 'Detailed information about package versions',
      code: 'conda search PKGNAME --info'
    },
    {
      desc: 'Remove unused cached files including unused packages',
      code: 'conda clean --all'
    },
    {
      desc: 'Remove a package from an environment',
      code: 'conda uninstall PKGNAME --name ENVNAME'
    },
    {
      desc: 'Update all packages within an environment',
      code: 'conda update --all --name ENVNAME'
    },
    {
      desc: 'Run most commands without requiring a user prompt. Useful for scripts.',
      code: 'conda install --yes PKG1 PKG2'
    },
    {
      desc: 'Examine Conda configuration',
      code: 'conda config --show'
    },
    {
      desc: 'Examine Conda configuration services',
      code: 'conda config --show-sources'
    }
  ]
}]

// mutable
sections = sections.map(section => {
  return {
    ...section,
    href: section.name.replace(/\s+/g, '-').toLowerCase()
  }
})

export const SECTIONS = sections


tool/compileWebVie.js
const pug = require('pug')
const fs = require('fs')

let sections = [{
  name: 'Getting Started',
  commands: [
    {
      desc: 'Verify Conda is installed, check version number',
      code: 'conda info'
    },
    {
      desc: 'Update Conda to the current version',
      code: 'conda update -n base conda'
    },
    {
      desc: 'Update all packages to teh latest version of Anaconda. Will install stable and compatible versions, not necessarily the very latest.',
      code: 'conda update anaconda'
    }
  ]
}, {
  name: 'Working with Environments',
  commands: [
    {
      desc: 'Create a new environment named ENVNAME with specific version of Python and packages installed.',
      code: 'conda create --name <ENVNAME> python=3.6 "PKG1>7.6" PKG2'
    },
    {
      desc: 'Activate a named Conda environment',
      code: 'conda activate ENVNAME'
    },
    {
      desc: 'Activate a Conda environment at particular location on disk',
      code: 'conda activate /path/to/environment-dir'
    },
    {
      desc: 'Deactivate current environment',
      code: 'conda deactivate'
    },
    {
      desc: 'List all packages and versions in the active environment',
      code: 'conda list'
    },
    {
      desc: 'List all packages and versions in the named environment',
      code: 'conda list --name ENVNAME'
    },
    {
      desc: 'List all revisions made within the active environment',
      code: 'conda list --revisions'
    },
    {
      desc: 'List all revisions made in a specified environment',
      code: 'conda list --name ENVNAME --revisions'
    },
    {
      desc: 'Delete an entire environment',
      code: 'conda remove --name ENVNAME --all'
    }
  ]
}, {
  name: 'Sharing Environments',
  commands: [
    {
      desc: 'Make an exact copy of an environment',
      code: 'conda create --clone ENVNAME --name NEWENV'
    },
    {
      desc: 'Export an environment to a YAML file that can be read on Windows, macOS, and Linux',
      code: 'conda env export --name ENVNAME > envname.yml'
    },
    {
      desc: 'Create an environment from YAML file',
      code: 'conda env create --file envname.yml'
    },
    {
      desc: 'Create an environment from the file named environment.yml in the current directory',
      code: 'conda env create'
    },
    {
      desc: 'Export an environment with exact package version for one OS',
      code: 'conda list --explicit > pkgs.txt'
    },
    {
      desc: 'Create an environment based on exact package versions',
      code: 'conda create --name NEWENV --file pkgs.txt'
    }
  ]
}, {
  name: 'Using Packages and Channels',
  commands: [
    {
      desc: 'Search for a package in currently configured channels with version range >=3.1.0, <3.2"',
      code: 'conda search PKGNAME=3.1 "PKGNAME [version=\'>=3.1.0,<3.2\']"'
    },
    {
      desc: 'Find a package on all channels using the Anaconda Client',
      code: 'anaconda search FUZZYNAME'
    },
    {
      desc: 'Install package from a specific channel',
      code: 'conda install conda-forge::PKGNAME'
    },
    {
      desc: 'Install a package by exact version number (3.1.4)',
      code: 'conda install PKGNAME==3.1.4'
    },
    {
      desc: 'Install one of the listed versions (OR)',
      code: 'conda install "PKGNAME[version=\'3.1.2|3.1.4\']"'
    },
    {
      desc: 'Install following several constraints (AND)',
      code: 'conda install "PKGNAME>2.5,<3.2"'
    },
    {
      desc: 'Add a channel to your Conda configuration',
      code: 'conda config --add channels CHANNELNAME'
    }
  ]
}, {
  name: 'Additional Useful Hints',
  commands: [
    {
      desc: 'Detailed information about package versions',
      code: 'conda search PKGNAME --info'
    },
    {
      desc: 'Remove unused cached files including unused packages',
      code: 'conda clean --all'
    },
    {
      desc: 'Remove a package from an environment',
      code: 'conda uninstall PKGNAME --name ENVNAME'
    },
    {
      desc: 'Update all packages within an environment',
      code: 'conda update --all --name ENVNAME'
    },
    {
      desc: 'Run most commands without requiring a user prompt. Useful for scripts.',
      code: 'conda install --yes PKG1 PKG2'
    },
    {
      desc: 'Examine Conda configuration',
      code: 'conda config --show'
    },
    {
      desc: 'Examine Conda configuration services',
      code: 'conda config --show-sources'
    }
  ]
}]

// mutable
sections = sections.map(section => {
  return {
    ...section,
    href: section.name.replace(/\s+/g, '-').toLowerCase()
  }
})

const compiledFunction = pug.compileFile('./tool/template.pug', {
  pretty: true
})
const html = compiledFunction({
  title: 'Conda Cheatsheet',
  sections
})

// save
const outputFilePath = './asset/index.html'
fs.writeFileSync(outputFilePath, html)
console.log(`Compile to: ${outputFilePath}`)


src/extension.ts
import * as vscode from 'vscode'
import * as path from 'path'
import * as fs from 'fs'

const CONDA_CHEATSHEET_URL: string = 'https://docs.conda.io/projects/conda/en/latest/user-guide/cheatsheet.html'

function getPdfDisposable (
  context: vscode.ExtensionContext,
  fileNames: string[],
  name: string,
  displayName: string
) {
  // set panel
  const assetPath = path.join(context.extensionPath, 'asset')

  const panel = vscode.window.createWebviewPanel(
    name,
    displayName,
    vscode.ViewColumn.Beside, {
      localResourceRoots: [vscode.Uri.file(assetPath)],
      enableScripts: false
    }
  )

  // set content
  let html: string = ''
  for (const fileName of fileNames) {
    const filePath = vscode.Uri.file(path.join(assetPath, fileName)).with({ scheme: 'vscode-resource' })
    html += `<img src="${filePath}" />`
  }
  panel.webview.html = html
}

export function activate (context: vscode.ExtensionContext) {
  console.log('Congratulations, your extension "conda-cheatsheet" is now active!')

  const pdfDisposable46 = vscode.commands.registerCommand('extension.openPdf46', () => {
    const fileNames = [
      'conda-4.6-1.jpg',
      'conda-4.6-2.jpg'
    ]
    const name = 'condaCheatsheetPdf46'
    const displayName = 'Conda Cheatsheet PDF (4.6)'
    getPdfDisposable(context, fileNames, name, displayName)
  })

  const pdfDisposable412 = vscode.commands.registerCommand('extension.openPdf412', () => {
    const fileNames = [
      'conda-4.12-1.jpg',
      'conda-4.12-2.jpg'
    ]
    const name = 'condaCheatsheetPdf412'
    const displayName = 'Conda Cheatsheet PDF (4.12)'
    getPdfDisposable(context, fileNames, name, displayName)
  })

  const pdfDisposable414 = vscode.commands.registerCommand('extension.openPdf414', () => {
    const fileNames = [
      'conda-4.14-1.jpg',
      'conda-4.14-2.jpg'
    ]
    const name = 'condaCheatsheetPdf414'
    const displayName = 'Conda Cheatsheet PDF (4.14)'
    getPdfDisposable(context, fileNames, name, displayName)
  })

  const pdfDisposable = vscode.commands.registerCommand('extension.openPdf', () => {
    const fileNames = [
      'conda-4.14-1.jpg',
      'conda-4.14-2.jpg'
    ]
    const name = 'condaCheatsheetPdf'
    const displayName = 'Conda Cheatsheet PDF (latest)'
    getPdfDisposable(context, fileNames, name, displayName)
  })

  const websiteDisposable = vscode.commands.registerCommand('extension.openWebsite', () => {
    // set panel
    const panel = vscode.window.createWebviewPanel(
      'condaCheatsheetWebsite',
      'Conda Cheatsheet Website',
      vscode.ViewColumn.Beside, {
        enableScripts: false
      }
    )

    // set content
    const html: string = `
      <style>html, body { height: 100%; !important }</style>
      <iframe style="position:relative; width: 100%; height: 100%; border: 0" src="${CONDA_CHEATSHEET_URL} "></iframe>
    `
    panel.webview.html = html
  })

  const webviewDisposable = vscode.commands.registerCommand('extension.openWebview', () => {
    // set panel
    const assetPath = path.join(context.extensionPath, 'asset')
    const panel = vscode.window.createWebviewPanel(
      'condaCheatsheetWebview',
      'Conda Cheatsheet Webview',
      vscode.ViewColumn.Beside, {
        localResourceRoots: [vscode.Uri.file(assetPath)],
        enableScripts: true
      }
    )

    // set content
    const styleFilePath = vscode.Uri.file(path.join(assetPath, 'custom.css')).with({ scheme: 'vscode-resource' })
    const scriptFilePath = vscode.Uri.file(path.join(assetPath, 'main.js')).with({ scheme: 'vscode-resource' })
    const htmlFilePath = path.join(assetPath, 'index.html')
    let html = fs.readFileSync(htmlFilePath, 'utf8')
    html = html.replace(/cspSource/g, panel.webview.cspSource.toString())
    html = html.replace('styleFilePath', styleFilePath.toString())
    html = html.replace('scriptFilePath', scriptFilePath.toString())
    panel.webview.html = html

    // listener, copy code
    panel.webview.onDidReceiveMessage(
      (code: string) => {
        vscode.env.clipboard.writeText(code)
          .then(() => {
            const message = `Copied: ${code}`
            vscode.window.setStatusBarMessage(message)
          }, () => {
            vscode.window.showErrorMessage('Copy to clipboard failed')
          })
      },
      undefined,
      context.subscriptions
    )
  })

  context.subscriptions.push(pdfDisposable46)
  context.subscriptions.push(pdfDisposable412)
  context.subscriptions.push(pdfDisposable414)
  context.subscriptions.push(pdfDisposable)
  context.subscriptions.push(websiteDisposable)
  context.subscriptions.push(webviewDisposable)
}

// this method is called when your extension is deactivated
export function deactivate () { }
